geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala

```diff
@@ -160,14 +164,135 @@ object QueryInterceptor extends LazyLogging {
+          throw new IllegalArgumentException(s"Graduated query guard configuration " +
+            s"has repeated size: ${first.sizeLimit}")
+        } else if (first.durationLimit.compareTo(second.durationLimit) <= 0) {
+          throw new IllegalArgumentException(s"Graduated query guard configuration " +
+            s"has durations out of order: ${first.durationLimit} is less than ${second.durationLimit}")
+        }
+      }
+      candidate
+    }
+
+    def buildLimits(guardConfig: util.List[_ <: Config]): Seq[SizeAndDuration] = {
+      import scala.collection.JavaConverters._
+      val confs = guardConfig.asScala.map { durationConfig =>
+        val size: Int = durationConfig.hasPath("size") match {
+          case true => durationConfig.getInt("size")
+          case false => Int.MaxValue
+        }
+        val duration = Duration(durationConfig.getString("duration"))
+        SizeAndDuration(size, duration)
+      }
+      evaluateLimits(confs)
+    }
+  }
+
+  class GraduatedQueryGuard extends QueryInterceptor {
+    var guardLimits: Seq[SizeAndDuration] = Seq()
+
+    /**
+     * Called exactly once after the interceptor is instantiated
+     *
+     * @param ds  data store
+     * @param sft simple feature type
+     */
+    override def init(ds: DataStore, sft: SimpleFeatureType): Unit = {
+      guardLimits = loadConfiguration(sft)
+    }
+
+    /**
+     * Modifies the query in place
+     *
+     * @param query query
+     */
+    override def rewrite(query: Query): Unit = { }
+
+    /**
+     * Hook to allow interception of a query after extracting the query values
+     *
+     * @param strategy query strategy
+     * @return an exception if the query should be stopped
+     */
+    override def guard(strategy: api.QueryStrategy): Option[IllegalArgumentException] = {
+      val intervalsOption: Option[FilterValues[Bounds[ZonedDateTime]]] = strategy.values.collect { case v: TemporalIndexValues => v.intervals }
+      val spatialBoundsOption: Option[Seq[(Double, Double, Double, Double)]] = strategy.values.collect { case v: SpatialIndexValues => v.spatialBounds }
+
+      val spatialExtent = spatialBoundsOption match {
+        case None => 360.0 * 180.0  // Whole world
+        case Some(seq) => seq.map { case  (lx: Double, ly: Double, ux: Double, uy: Double) =>
+          (ux-lx) * (uy-ly)
+        }.sum
+      }
+
+      intervalsOption match {
+        case None => // Q: What does None here mean again?!
+          None       // Returning None means that the query continues.  Is this correct?
+        case Some(intervals) =>
+          if (!intervals.forall(_.isBoundedBothSides) || intervals.isEmpty) {
+            Some(new IllegalArgumentException("At least one part of the query is temporally unbounded.  Add bounded temporal constraints."))
+          } else {
+            val queryDuration = duration(intervals.values)
+            guardLimits.find( _.sizeLimit > spatialExtent).flatMap {
```

Comment:
should this be `>=`? also, seems like there might not be a matching limit, in that case the query would go through, but seems like it shouldn't? since you would have configured a limit at a smaller size. (although also if the config is wrong the guards would be empty...)

Reply:
i was thinking `>=` would be needed to match full world scans

Reply:
You are correct.  I've added unit test cases to help with the corner cases.
The size of whole world guards is Max.Int:).